### PR TITLE
Feature: RPC

### PR DESCRIPTION
**Pull request by plsthink**
Open PR for discussion about implementing RPC service inside go-nitro lib.

Please check commits and examples code comments for further details.

### CODE REVIEW

File: rpc/service.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			}
+
+			go s.HandlePeer(peer)
+		}
+	}()
+
+	return s
+}
+
+func (s *Service) HandlePeer(p *network.Peer) {
```
Comment (calj) [inline rpc/service.go]: Peer could be named **NetworkServiceConnection** here
**HandleNetworkServiceConnection**

File: channel/state/outcome/exit.go [hunk 2]
```diff
@@ -13,9 +13,9 @@ import (
 
 // SingleAssetExit declares an ordered list of Allocations for a single asset.
 type SingleAssetExit struct {
-	Asset       types.Address // Either the zero address (implying the native token) or the address of an ERC20 contract
-	Metadata    []byte        // Can be used to encode arbitrary additional information that applies to all allocations.
-	Allocations Allocations
+	Asset       types.Address `json:"asset"`    // Either the zero address (implying the native token) or the address of an ERC20 contract
+	Metadata    []byte        `json:"metadata"` // Can be used to encode arbitrary additional information that applies to all allocations.
+	Allocations Allocations   `json:"allocations"`
```
Comment (mod) [inline channel/state/outcome/exit.go:18]: We are not suppose to transport this structure. or maybe ObjectiveRequest already do it somehow?

File: client/client.go [hunk 3]
```diff
@@ -22,12 +22,12 @@ import (
 
 // Client provides the interface for the consuming application
 type Client struct {
```
Comment (mod) [inline client/client.go:24]: if those changes are not mandatory for our PR, I suggest we discard using git checkout

File: examples/rpc_nats/main.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	// for exemple: disconnect from B or I, reconnect to B or I, ...
+	// Orverall, the goal is to be able to completly control the client trough the rpc service
+
+	// Setup B and I clients
+	chainServiceB := chainservice.NewMockChainService(chain, bob.Address())
+	chainServiceI := chainservice.NewMockChainService(chain, irene.Address())
+
+	clientA, msgA := setupClientWithP2PMessageService(alice.PrivateKey, 3005, chainServiceA, logger)
+	clientB, msgB := setupClientWithP2PMessageService(bob.PrivateKey, 3006, chainServiceB, logger)
+	clientI, msgI := setupClientWithP2PMessageService(irene.PrivateKey, 3007, chainServiceI, logger)
```
Comment (mod) [inline examples/rpc_nats/main.go:95]: I don't understand the name.

File: network/service_test.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+func newNetworkServiceMock() *NetworkService {
+	service := NetworkService{
+		Logger:              zerolog.Logger{},
+		Connection:          &connMock{},
+		Serde:               &serializerMock{},
+		handlerRequest:      sync.Map{},
+		handlerResponse:     sync.Map{},
+		handlerError:        sync.Map{},
+		handlerPublicEvent:  sync.Map{},
+		handlerPrivateEvent: sync.Map{},
```
Comment (mod) [inline network/service_test.go:55]: Are we sure we should do so many maps ?
can't we just do Request, Response ?

File: protocols/directfund/directfund.go [hunk 6]
```diff
@@ -445,12 +445,12 @@ func IsDirectFundObjective(id protocols.ObjectiveId) bool {
...
-	AppDefinition     types.Address
-	AppData           types.Bytes
-	Nonce             uint64
+	CounterParty      types.Address `json:"counter_party"`
+	ChallengeDuration uint32        `json:"challenge_duration"`
+	Outcome           outcome.Exit  `json:"outcome"`
+	AppDefinition     types.Address `json:"app_definition"`
+	AppData           types.Bytes   `json:"app_data"`
+	Nonce             uint64        `json:"nonce"`
 }
```
Comment (mod) [inline protocols/directfund/directfund.go:454]: did they had serde already ?
in which case do we want to send and receive ObjectiveRequests

**Review by mod — CHANGES_REQUESTED** [review]
I don't fully understand why this is so complicated.

Why interfere with objectives? this is the internal kitchen of nitro2nitro.

I was thinking we expose a simplified API, with a simple handler,
we parse params, and provide the protocols requested with the params.

File: channel/state/outcome/exit.go (hunk 2 above)
Comment (calj) [reply channel/state/outcome/exit.go:18]: this is something to cleanup

**Pull request closed without merging**